cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -221,75 +277,72 @@ private void processMetrics() {
         MessageFetcher fetcher = messagingService.prepareFetch(topicIdMetaKey.getTopicId());
         fetcher.setLimit(fetcherLimit);
         if (lastMessageId != null) {
+          LOG.debug("Start fetching from lastMessageId = {}", lastMessageId);
           fetcher.setStartMessage(lastMessageId, false);
         } else {
+          LOG.debug("Start fetching from beginning");
           fetcher.setStartTime(0L);
         }
 
-        byte[] nextMessageId = null;
+        byte[] currentMessageId = null;
         try (CloseableIterator<RawMessage> iterator = fetcher.fetch()) {
           while (iterator.hasNext() && isRunning()) {
             RawMessage input = iterator.next();
             try {
               payloadInput.reset(input.getPayload());
               MetricValues metricValues = recordReader.read(decoder, recordSchema);
               records.add(metricValues);
-              nextMessageId = input.getId();
+              currentMessageId = input.getId();
               LOG.trace("Received message {} with metrics: {}", Bytes.toStringBinary(lastMessageId), metricValues);
             } catch (IOException e) {
               LOG.warn("Failed to decode message to MetricValue. Skipped. {}", e.getMessage());
             }
           }
         }
 
+        if (currentMessageId != null) {
+          topicMessageIds.put(topicIdMetaKey, currentMessageId);
+          lastMessageId = currentMessageId;
+        }
+
+        // Skip persisting if the records is empty and the lastMessageId has already been persisted in the meta table
         if (records.isEmpty()) {
-          LOG.trace("No metrics record to process.");
-          return;
+          if (lastMessageId != null && Arrays.equals(lastPersistedMessageIdMap.get(topicIdMetaKey), lastMessageId)) {
+            return;
+          }
         }
 
-        persistRecords(records);
-        records.clear();
-        lastMessageId = nextMessageId;
-        persistMessageId(lastMessageId);
+        // Ensure there's only one thread can persist records when the persistingFlag is false.
+        // The thread set persistingFlag to true when it starts to persist.
+        if (!persistingFlag.compareAndSet(false, true)) {
+          LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
+          return;
+        }
+        try {
+          Map<TopicIdMetaKey, byte[]> topicMessageIdsCopy = new HashMap<>(topicMessageIds);
+          LinkedList<MetricValues> recordsCopy = new LinkedList<>();
+          while (!records.isEmpty()) {
+            recordsCopy.add(records.poll());
+          }
+          if (!recordsCopy.isEmpty()) {
+            persistRecords(recordsCopy);
+          }
+          persistMessageIds(topicMessageIdsCopy);
```

Comment:
Failing in persisting the topic message ids is ok, right? As it will get retried in the next fetch iteration. We can do an extra persist on the service shutdown.